File: c/leetcode/需要二刷的题/234 回文链表/another.c
//原来如此！我悟了！原本就想过O(1)的时间肯定是从中间反转，然后头和中间同时开始遍历链表
//当时觉得难点是如何找到中间的节点，a->val == a->next->val肯定不可行，然后就陷入思维僵局了……
//怎么就忘了快慢指针啊……可恶啊

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */


bool isPalindrome(struct ListNode* head){
    struct ListNode *fast=head;
    struct ListNode *before = NULL;
    struct ListNode *after=head;
    while(fast->next && fast->next->next){
        fast = fast->next->next;
        after = head->next;
        head->next = before;
        before = head;
        head = after;
    }
    if(fast == head){//fast == head 的情况只会在链表节点小于等于两个的时候才会发生，否则fast每次走两步，head走一步，肯定不会相同
        if(head->next == NULL)//只有一个元素的情况下，肯定是回文的
            return true;
        else if(head->next->val == head->val)//两个元素的情况下，如果相同则是回文的
            return true;
        else
            return false;
    }
    after = head->next;
    head->next = before;
    if(fast->next != NULL){
        fast = after;
    }
    else{
        head = head->next;
        fast = after;
    }
    while(head){
        if(head->val != fast->val)
            return false;
        head = head->next;
        fast = fast->next;
    }
    return true;
}
//写是可以写的没什么问题，就是这写的好像有点累啊……五十行的代码，说明写的绝不轻松，中间从25-41这十六行
//都是基于样例的条件判断，根据样例写代码实锤了，说明写之前并没有怎么规划好

File: leetcode/需要二刷的题/92 反转链表/another.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseBetween(head *ListNode, left int, right int) *ListNode {
	if left == right {
		return head
	}
	res := &ListNode{Val: 0, Next: head}
	head = res
	var prev *ListNode = nil
	var after *ListNode = nil
	for i := 0; i < left-1; i++ {
		head = head.Next
	}
	prev = head
	head = head.Next
	for i := 0; i < right-left; i++ {
		after = head.Next
		head.Next = after.Next
		after.Next = prev.Next
		prev.Next = after
	}
	return res.Next
}

//=.=头插法 想着练一下，妹想到一练练了四十分钟，真是个傻逼
//真的，下次别头铁了，求你了。链表题老老实实画图做，别脑子硬想，想不过来的，又不是什么难题，画图三五分钟解决的事情，硬要在这里头铁，有毒……